### PR TITLE
xieyu_wang/task_05

### PR DESCRIPTION
**Pull request by AncientEncoder**

### CODE REVIEW

**Review by Friend-LGA — CHANGES_REQUESTED** [review]
Please, put the label

File: xieyu_wang/Logger.hpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  }
+  void setFile(const std::optional<std::string>& filename);
+  void log(const std::string& string);
+  ~Logger();
+
+ private:
+  std::optional<std::ofstream> file_;
+  std::mutex mutex_;
+
+  Logger() = default;
```
Comment (Friend-LGA) [inline xieyu_wang/Logger.hpp:22]: Again, did you read the task?
You need to add `config.hpp`, and take path to log file from there.
```suggestion
  Logger() : file_(config::LOGGER_FILE_PATH) {
    if (!file_.is_open()) {
      throw std::runtime_error("Failed to open log file");
    }
  }
```
If you do like this, you don't need any `optional` and you don't need `setFile`

File: xieyu_wang/main.cpp [hunk 2]
```diff
@@ -1,9 +1,79 @@
...
+      std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  std::stringstream TimeInString;
+  TimeInString << std::put_time(std::localtime(&presentTime), "%d/%m/%Y %T");
+  return TimeInString.str();
+}
+void logStart(Logger& logger, int graphIndex) {
+  logger.log(getTimeByString() + ": Graph " + std::to_string(graphIndex) +
+             ", Generation Started\n");
+}
+int getMaxDepth(Logger& logger, Graph graph) {
```
Comment (Friend-LGA) [inline xieyu_wang/main.cpp:26]: Why logger is here?

File: xieyu_wang/main.cpp [hunk 3]
```diff
@@ -1,9 +1,79 @@
...
+void logEnd(Logger& logger, Graph graph, int graphIndex) {
+  logger.log(getTimeByString() + ": Graph " + std::to_string(graphIndex) +
+             ", Generation Finished {  \n");
+  logger.log("  depth: " + std::to_string(getMaxDepth(logger, graph)) + ",\n");
+  logger.log("  vertices: " + std::to_string(graph.getVertices().size()) +
+             ", [");
+  logDepth(logger, graph, getMaxDepth(logger, graph));
+  logger.log("],\n  edges: " + std::to_string(graph.getEdges().size()) + ", {");
+  logColors(logger, graph);
+  logger.log("}\n}\n");
```
Comment (Friend-LGA) [inline xieyu_wang/main.cpp:65]: All this into `graph_printing.hpp/cpp`
```cpp
namespace graph_printing {
std::string print_graph_description(const Graph& graph);
}
```

File: xieyu_wang/Edge.hpp [hunk 4]
```diff
@@ -14,3 +15,5 @@ class Edge {
   std::string toString() const;
```
Comment (Friend-LGA) [inline xieyu_wang/Edge.hpp:15]: Everything related to printing should be moved to `graph_printing.hpp/cpp`
```cpp
namespace graph_printing {
std::string print_graph(const Graph& graph);
std::string print_vertex(const Vertex& vertex);
std::string print_edge(const Edge& edge);
std::string print_edge_color(const Edge::Color& color);
}
```

File: xieyu_wang/main.cpp [hunk 5]
```diff
@@ -1,9 +1,79 @@
...
+}
+Logger& setUpLogger() {
+  auto& logger = Logger ::getLogger();
+  logger.setFile("./temp/log.txt");
+  return logger;
+}
 int main() {
```
Comment (Friend-LGA) [inline xieyu_wang/main.cpp:72]: function is too long. Split it into smaller functions.
Task description has an [example](https://github.com/Friend-LGA/uni_course_cpp/tree/master/practice/task_05#%D1%84%D1%83%D0%BD%D0%BA%D1%86%D0%B8%D1%8F-main-%D0%B2%D0%B0%D1%88%D0%B5%D0%B9-%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D1%8B). Did you even read it?

**Review by Friend-LGA — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 6]
```diff
@@ -46,3 +46,4 @@ shamil_latypov/temp
 /kirill_tolstobrov/temp/
 /nikolai_chernyshov/temp/
 /robert_gaisin/temp/
+/xieyu_wang/temp/
```
[the remark this answers is not in the document]
Comment (AncientEncoder) [reply .gitignore:49]: sry shouldn’t I change here ?

**Review by Friend-LGA — CHANGES_REQUESTED** [review]

File: xieyu_wang/Graph.cpp [hunk 7]
```diff
@@ -107,17 +87,17 @@ const Edge& Graph::getEdge(int id) {
   throw std::runtime_error("No such vertex");
 }
 
-int Graph::getNumEdgeByColor(const Edge::Color color) {
-  int sum = 0;
+std::vector<Edge> Graph::getNumEdgeByColor(const Edge::Color& color) {
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:90]:
```cpp
getEdgesByColor
```
return by reference

File: xieyu_wang/Graph.cpp [hunk 8]
```diff
@@ -107,17 +87,17 @@ const Edge& Graph::getEdge(int id) {
...
+  std::vector<Edge> colorEdges;
   for (const auto& edge : edges_) {
     if (edge.color == color) {
-      sum++;
+      colorEdges.push_back(edge);
     }
   }
-  return sum;
+  return colorEdges;
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:97]: add to graph some map of colored edges like
```cpp
std::unordered_map<Color, std::vector<EdgeId>> colored_edges_;
```
```cpp
if (colored_edges_.find(color) == colored_edges_.end()) {
  static std::vector<EdgeId> empty_result = {};
  return empty_result;
}
return colored_edges_.at(color);
```

File: xieyu_wang/Graph.hpp [hunk 9]
```diff
@@ -7,12 +7,11 @@ class Graph {
   const std::vector<Vertex>& getVertices() const;
   const Vertex& addVertex();
   void addEdge(int fromVertexId, int toVertexId, const Edge::Color color);
-  std::string toString() const;
   bool hasVertex(int idFind);
   bool isConnected(int fromVertexId, int toVertexId);
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.hpp:11]: const method

File: xieyu_wang/Graph.hpp [hunk 10]
```diff
@@ -7,12 +7,11 @@ class Graph {
   const std::vector<Vertex>& getVertices() const;
   const Vertex& addVertex();
   void addEdge(int fromVertexId, int toVertexId, const Edge::Color color);
-  std::string toString() const;
   bool hasVertex(int idFind);
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.hpp:10]: const method

File: xieyu_wang/GraphPrinter.hpp [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#include "Edge.hpp"
+#include "Graph.hpp"
+
+using Graph = uni_course_cpp::Graph;
+using Edge = uni_course_cpp::Edge;
+using Vertex = uni_course_cpp::Vertex;
```
Comment (Friend-LGA) [inline xieyu_wang/GraphPrinter.hpp:6]: We want to avoid using of `using` inside `hpp` file

**Review by Friend-LGA — CHANGES_REQUESTED** [review]

File: xieyu_wang/Graph.cpp [hunk 12]
```diff
@@ -87,14 +88,12 @@ const Edge& Graph::getEdge(int id) {
   throw std::runtime_error("No such vertex");
 }
 
-std::vector<Edge> Graph::getNumEdgeByColor(const Edge::Color& color) {
-  std::vector<Edge> colorEdges;
-  for (const auto& edge : edges_) {
-    if (edge.color == color) {
-      colorEdges.push_back(edge);
-    }
+std::vector<Edge>& Graph::getEdgeByColor(const Edge::Color& color) {
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:91]:
```suggestion
const std::vector<int>& Graph::getEdgesByColor(const Edge::Color& color) const {
```

File: xieyu_wang/Graph.hpp [hunk 13]
```diff
@@ -22,5 +23,6 @@ class Graph {
   int edgeIdCounter_ = 0;
   std::vector<Vertex> vertices_;
   std::vector<Edge> edges_;
+  std::unordered_map<Edge::Color, std::vector<int>> coloredEdges_;
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.hpp:26]: `EdgeId`

File: xieyu_wang/Graph.hpp [hunk 14]
```diff
@@ -7,10 +8,10 @@ class Graph {
   const std::vector<Vertex>& getVertices() const;
   const Vertex& addVertex();
   void addEdge(int fromVertexId, int toVertexId, const Edge::Color color);
-  bool hasVertex(int idFind);
-  bool isConnected(int fromVertexId, int toVertexId);
+  const bool hasVertex(int idFind);
+  const bool isConnected(int fromVertexId, int toVertexId);
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.hpp:12]: don't return const copy. Make the method const

File: xieyu_wang/LoggingHelper.cpp [hunk 15]
```diff
@@ -40,8 +40,7 @@ void LoggingHelper::logColors(Logger& logger, Graph graph) {
       Edge::Color::Yellow, Edge::Color::Red};
   for (int i = 0; i < colors.size(); i++) {
```
Comment (Friend-LGA) [inline xieyu_wang/LoggingHelper.cpp:41]:
```cpp
for (const auto& color : colors) {
```

**Review by Friend-LGA — CHANGES_REQUESTED** [review]

File: xieyu_wang/Graph.cpp [hunk 16]
```diff
@@ -55,11 +55,11 @@ const bool Graph::isConnected(int fromVertexId, int toVertexId) {
   return false;
 }
 
-Vertex& Graph::getVertex(int id) {
+Vertex& Graph::getVertex(int id) const {
   assert(hasVertex(id) && "Vertex doesn't exist");
   for (auto& vertex : vertices_) {
     if (vertex.id == id) {
-      return vertex;
+      return (Vertex&)vertex;
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:62]: This type casting should not be here

File: xieyu_wang/Graph.hpp [hunk 17]
```diff
@@ -8,15 +8,15 @@ class Graph {
...
+  bool isConnected(int fromVertexId, int toVertexId) const;
   std::vector<int> getVertexIdsAtDepth(int depth);
-  std::vector<Edge>& getEdgeByColor(const Edge::Color& color);
+  const std::vector<int>& getEdgesByColor(const Edge::Color& color) const;
   const std::vector<Edge>& getEdges() const;
 
  private:
-  Vertex& getVertex(int id);
-  const Edge& getEdge(int id);
+  Vertex& getVertex(int id) const;
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.hpp:18]: If you need non-const then don't make it const

**Review by Friend-LGA — CHANGES_REQUESTED** [review]

File: xieyu_wang/Graph.cpp [hunk 18]
```diff
@@ -24,15 +24,15 @@ void Graph::addEdge(int fromVertexId, int toVertexId, const Edge::Color color) {
     toVertex.depth = fromVertex.depth + 1;
   }
 }
-bool Graph::hasVertex(int idFind) const {
+bool Graph::hasVertex(int idFind) {
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:27]: const method

File: xieyu_wang/Graph.cpp [hunk 19]
```diff
@@ -24,15 +24,15 @@ void Graph::addEdge(int fromVertexId, int toVertexId, const Edge::Color color) {
...
   for (const auto& vertex : vertices_) {
     if (vertex.id == idFind) {
       return true;
     }
   }
   return false;
 }
-bool Graph::isConnected(int fromVertexId, int toVertexId) const {
+bool Graph::isConnected(int fromVertexId, int toVertexId) {
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:35]: const method

File: xieyu_wang/Graph.cpp [hunk 20]
```diff
@@ -55,11 +55,11 @@ bool Graph::isConnected(int fromVertexId, int toVertexId) const {
   return false;
 }
 
-Vertex& Graph::getVertex(int id) const {
+Vertex& Graph::getVertex(int id) {
```
Comment (Friend-LGA) [inline xieyu_wang/Graph.cpp:58]: const version of this method if you need so

**Review by Friend-LGA — CHANGES_REQUESTED** [review]
Mak

**Review by Friend-LGA — APPROVED** [review]
🎉

**Pull request merged**